proposals/mdm-baseline.md

```diff
@@ -0,0 +1,153 @@
+# MDM core
+
+This document is a high-level spec/guide of the technical challenges that don't
+directly map to a product feature request.
+
+Some of this might already be captured in other issues, but we wanted to put
+everything together and have multiple people thinking about the solutions as a
+whole.
+
+We will create issues to tackle actionable bits, but the idea is to also use
+this to guide the implementation of feature requests.
+
+
+### Problem: Hosts can be enrolled via MDM
+
+Currently, there's no reliable way to associate an MDM enrolled device with a
+host in the `hosts` table.
+
+- If a host is enrolled in MDM before its enrolled via `osquery` (eg: DEP
+  enrollment) we will have a record in the `nano_enrollments` table before the
+  host exists in the `hosts` table.
+
+- If a host exists in both tables, we could `JOIN` them using `hosts.uuid =
+  nano_enrollments.device_id`  but we need to make sure the right constraints
+  in the application logic exist and the right indexes are created for fast
+  access.
+
+Currently, the code thinks of these two tables as two disjoint sets of elements
+that represent the same entity: a host.
+
+Motivation:
+
+- In https://github.com/fleetdm/fleet/issues/8878 we're asked to return hosts
+  that are scheduled for DEP enrollment (not even enrolled yet) to the `GET
+  /hosts` response.
+- In https://github.com/fleetdm/fleet/issues/8878 we're asked to retrieve a
+  list of hosts enrolled via `osquery` that are pending to be enrolled via MDM.
+
+#### Proposed solution
+
+`hosts` should always be the source of truth, we could consider
+`nano_enrollments` the same way we treat  the "auxiliary" `hosts_*` tables (eg:
+`host_disks`): as tables holding metadata for a host. 
+
+When we get information through MDM channels (be it via an enrollment, or via
+the DEP API notifying a new device purchased,) we always get a device serial
+number.
+
+1. Ensure there's a host in the `hosts` table with that serial number (insert
+   one if there isn't.) Try to pre-populate as much info as possible from the
+   MDM payload (eg: `ProductName` is `hosts.hardware_model`)
+1. Create a new entry in the related `nano_*` table pointing to the host.
+
+This implies that we must add an uniqueness constraint for the serial number in both tables.
```

Comment:
Hm, I wonder if this might be in conflict with VMs being enrolled in some cases. We should check with the agent folks before adding a constraint in the hosts table, that sounds like a problematic migration in many ways.

Reply:
Agreed on possible conflicts with VMs. I don't think we would want to enforce uniqueness through a DB constraint because we are inevitably going to see devices (particularly VMs, IoT devices, etc.) that have weird/missing/duplicated serials. For macOS hosts it seems more reasonable to expect this, though I wonder if there are virtualization scenarios where that assumption might still be broken? Is it possible for us to use the serial to associate the enrollments for MDM purposes but still make it possible to support separate entries for non-MDM devices (or something like this)?

Reply:
gotcha, that makes sense to me.

>  Is it possible for us to use the serial to associate the enrollments for MDM purposes but still make it possible to support separate entries for non-MDM devices (or something like this)?

I think so, we'll need to do something about the possibility of having more that one row with the same serial number, eg. @chiiph proposed in another comment (and makes sense to me)

>  if we DON'T add the constraint in the hosts table, we should only have 1 unenrolled host from DEP with that serial number (because it's new hardware... how would it have been duplicated?) so what if we just grab the first one we get from the hosts table and log if there are more.

cc: @michalnicp @gillespi314